[REST API] Tracking

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8190
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds tracking to the site credentials login and application passwords usage according to this plan pe5sF9-13h-p2 

For some tracking events, I needed two preparatory changes, and this what made the PR bigger:
1. [57af766](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/57af7660e2a9791f66c8afb264b116d0f441af16): to expose the error when fetching user role fails.
2. [e9ebed7](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/e9ebed7b610df2154e2dfae7e0da86833cd76a27): to expose events about application password generation failures.

For the rest, since some events were added in previous PRs, I prepared this table to help you follow where each event was added:

| Events | Commit |
| --- | --- |
| login_site_address_site_info_requested <br /> login_site_address_site_info_succeeded | 812badeec85fbb0f24a7ed86661b3dfca92be859 |
| unified_login_step (login_store_creds) <br/> unified_login_interaction (submit)  | 67885531fd878f7b7cc3ade537640858c2925b15 (previous PR) |
| unified_login_failure <br /> login_site_credentials_login_failed | 306b8d76f935ed028ba5f46a49db0c965d6eeb64 |
| unified_login_step (not_woo_store) |  b57fbaef54c012c333e64813f1a6b04f5f11bddd (previous PR) |
| login_insufficient_role | f6547ff107cdd12bb1f2994c274554ef6ebf839c |
| rest_api_login_experiment | 87475d00cfc2daac54766cffabe74b7bd88b2564 (previous PR) |
| signed_in |  ac3c712517c69ee8eec7a391572dc5e2469bbe5f |
| application_passwords_new_password_created <br/> application_passwords_generation_failed | b984f38094f813cdbd584f0d325cd05162468671 |

### Testing instructions
I tried to have detailed test scenarios below because they helped me validate the events, I don't think it's expected to test all those scenarios in a code review, just test what you can, then review the code and check the tests added in [88a2d3a](https://github.com/woocommerce/woocommerce-android/pull/8220/commits/88a2d3a1656f50ab27413c878a0ac1a6bb3f1ca2)

To test error scenarios below, it's better to use a proxy app (like Charles), and you might find a website that simulates different status codes like https://httpstat.us/ useful:

1. Make sure you are assigned the Treatment variant of the REST API experiment (either by [hardcoding](https://github.com/woocommerce/woocommerce-android/blob/9841083ed877c1846ca66d43a44841ecfb320837/WooCommerce/src/main/kotlin/com/woocommerce/android/experiment/RESTAPILoginExperiment.kt#L43) it or by setting a test device)
2. Open the app, then click on "Enter site address"
3. Enter the site address.
4. In the next screen, enter wrong site credentials.
5. Confirm the following events are sent:
```
unified_login_interaction, Properties: {"flow":"login_store_creds","step":"username_password", "click":"submit"...}
unified_login_failure, Properties: {"flow":"login_store_creds","step":"username_password" ...}
login_site_credentials_login_failed, Properties: { "step":"authentication" ... }
```
6. Simulate a site that has Application Passwords disabled (`wp/v2/users/me/application-passwords` should return an error that contains the returns the body: `{ "code" : "application_passwords_disabled" }`), or alternatively install a plugin that disable them.
7. Attempt login again but using correct credentials.
8. Confirm the following event is sent: `application_passwords_generation_failed, Properties: {"cause":"feature_disabled","scenario":"generation",...}
9. Re-enable application passwords.
10. Simulate a generic error for `wp/v2/users/me/application-passwords`
11. Attempt login again.
12. Confirm the following events are sent:
```
application_passwords_generation_failed, Properties: {"cause":"other","scenario":"generation"...}
login_site_credentials_login_failed, Properties: {"step":"application_password_generation"...}
```
13. Remove the error of `wp/v2/users/me/application-passwords`, then simulate a generic error for `/wc/v3/settings`
14. Attempt login again.
15. Confirm the following event are sent: 
```
application_passwords_new_password_created: Properties: { "scenario":"generation"...}
login_site_credentials_login_failed, Properties: {"step":"woo_status"...}
```
16. Remove the error of `/wc/v3/settings`, then simulate a generic error for `/wp/v2/users/me`
17. Confirm the following event is sent: `login_site_credentials_login_failed, Properties: {"step":"user_role"...}`
18. Remove the error simulation.
19. Attempt login.
20. Confirm the following event is sent: `signed_in`
21. Go to your site, then change your username's password and revoke the application's password.
22. Trigger any API requests in the app.
23. Confirm the following event is sent: `application_passwords_generation_failed, Properties: {"cause":"authorization_failed","scenario":"regeneration"...}`
 

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Update site info API events

hichamboushaba committed:
Update UserEligibilityFetcher to expose error details

hichamboushaba committed:
Expose password generation failures

hichamboushaba committed:
Add error tracking to site credentials login

hichamboushaba committed:
Track user role failure

hichamboushaba committed:
Add application passwords tracking events

hichamboushaba committed:
Fix unit tests

hichamboushaba committed:
Move signed in tracking after fetching user role

hichamboushaba committed:
Add tests for the analytics events

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/UserEligibilityErrorViewModel.kt:
```diff
@@ -20,18 +23,29 @@ import javax.inject.Inject
+            analyticsTracker.track(
+                AnalyticsEvent.LOGIN_INSUFFICIENT_ROLE,
+                mapOf(
+                    ROLES_KEY to user?.getUserRoles()?.joinToString(",") { it.value }
```
There's an issue related to this tracking although it's not caused by the code changes from this PR, but IMO it would make sense to address it as part of the new tracking for store creds login. 
This tracking will sometimes track an empty string for the `ROLE_KEY` value:
```
login_insufficient_role, Properties: {"current_roles":",,","is_debug":true}
```
This is because some roles like `contributor` are mapped to `WCUserRole.OTHER`. And `OTHER` roles has an empty String as `value` so in this line: 
```kotlin 
ROLES_KEY to user?.getUserRoles()?.joinToString(",") { it.value }
```
`it.value` is ""

I think we should add a `value = "other"` to `WCUserRole.OTHER`, or even better map all the existing roles from `wc-admin` to `WCUserRole`

hichamboushaba replied:
I fixed this in 14901aa1606a94ce80d4e41df32d36441c98089a, the fix is a bit big, I had the choice of either going to FluxC and doing the fix there, or using the domain class `User` that we had in our codebase (I just noticed it). I went with the second approach because IMO it's better to have the logic in the domain layer.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:
```diff
@@ -134,49 +141,75 @@ class LoginSiteCredentialsViewModel @Inject constructor(
-            triggerEvent(ShowSnackbar(R.string.error_generic))
-        }
+        userEligibilityFetcher.fetchUserInfo().fold(
+            onSuccess = {
+                loginAnalyticsListener.trackAnalyticsSignIn(false)
+                triggerEvent(LoggedIn(selectedSite.getSelectedSiteId()))
+            },
+            onFailure = { exception ->
```
Double-checking if this is expected in regards considering the following scenario as a successful `signed_in`. Imho it shouldn't. 
At this point in the code the following tracking will be sent: 
`Tracked: signed_in, Properties: {"dotcom_user":false,"is_debug":true}` 
Even in cases where user has an invalid/insufficient role to log in and thus lands in `UserEligibilityErrorFragment`. This code also won't track `login_site_credentials_login_failed` -> `user_role` even though the user has an invalid role.

hichamboushaba replied:
Nice catch, I did this to follow what we do with the WPCom login (we track login success before checking the user role check), but I agree it's better to avoid it for our case, I added a check to avoid sending the event in c49bad5.

> This code also won't track login_site_credentials_login_failed -> user_role even though the user has an invalid role.

This state is tracked using the other event `login_insufficient_role`, the `login_failed` events are for failures caused by API or network issues.

JorgeMucientes commented:
Excellent job @hichamboushaba. 
I tested most of the scenarios except for those that required proxying to return a generic response error. The rest of the scenarios that could be set up by changing config in `wp-admin` I've tested all them and worked well. 
I did notice a couple of issues that, I think, are worth looking into before merging. Check my 2 previous comments.

hichamboushaba committed:
Add unit tests

hichamboushaba committed:
Improve the User class

The commit also updates UserEligibilityFetcher to use the app model instead of the DB one

hichamboushaba committed:
Emit the signed in success event only if the user is eligible

hichamboushaba committed:
Fix tests

hichamboushaba commented:
Thanks @JorgeMucientes for the review, ready for another round.

hichamboushaba committed:
Fix logic of signing when user is not eligible

hichamboushaba committed:
Fix display of user roles in the error screen

hichamboushaba committed:
Fix logic of parsing user roles

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/User.kt:
```diff
@@ -35,6 +41,90 @@ fun WCUserModel.toAppModel(): User {
+sealed class UserRole(val value: String) : Parcelable {
+    companion object {
+        fun fromString(role: String): UserRole {
+            return when (role) {
+                "owner" -> Owner
+                "administrator" -> Administrator
+                "shop_manager" -> ShopManager
+                "editor" -> Editor
+                "author" -> Author
+                "customer" -> Customer
+                "subscriber" -> Subscriber
+                else -> Other(role)
```
We can use the [`sealedSubclasses`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-class/sealed-subclasses.html) property to achieve this without the long `when`, but it requires the `kotlin-reflect` dependency, and I didn't want to add it just for this part.

JorgeMucientes commented:
Changes look great! Thanks for applying them 🙏🏼
